src/script.c

```diff
@@ -463,14 +466,18 @@ static int scriptVerifyClusterState(scriptRunCtx *run_ctx, client *c, client *or
      * already been thrown. This is only checking for cross slot keys being accessed
      * that weren't pre-declared. */
     if (hashslot != -1 && !(run_ctx->flags & SCRIPT_ALLOW_CROSS_SLOT)) {
-        if (original_c->slot == -1) {
-            original_c->slot = hashslot;
-        } else if (original_c->slot != hashslot) {
+        if (run_ctx->slot == -1) {
+            run_ctx->slot = hashslot;
+        } else if (run_ctx->slot != hashslot) {
             *err = sdsnew("Script attempted to access keys that do not hash to "
                     "the same slot");
             return C_ERR;
         }
     }
+
+    c->slot = hashslot;
+    original_c->slot = hashslot;
```

Comment:
don't we wanna do that only if `hashslot != -1`?

Reply:
I  we should it according to the slot being accessed by the current command, even if it is `-1`. For example, in the case of `redis.call('ping')`, we should handle each command in the script just like handling commands from regular network clients.

Reply:
I think it makes more sense that c->slot reflects both the intended and actual access patterns, and should be set to some invalid value on errors. This value is used primarily as a cached value, so not sure we want to be caching an incorrect value. So:
1. If client declares no slot, and accesses slot 0, c->slot is marked as slot 0.
2. If a client declares slot 0, and accesses slot 1, c->slot is marked as invalid.
3. If a client declares no slot, and accesses slot 0 and slot 1, c->slot is marked as invalid.

Reply:
Is `c->slot` referring to the script `engine_client` here? It sounds like a good suggestion, but it seems that we don't actually use it anywhere (especially in this PR, where I replaced `c->slot` with `getKeySlot()` in sort and pubsub). Can you provide an example?

Also, I am concerned about the potential risks of doing this. If we don't set the desired slot for the `engine_client->slot`, it could introduce some potential risks because, after all, its name is still 'slot,' which can easily lead to direct usage. For example, in the modified sort and pubsub commands.

Reply:
Is there still a problem with this?

Reply:
no, i think we're good

Reply:
Sorry for circling back slowly. You're right that it's not used, maybe I'll submit a PR to rename it as "declared slot", which might diverge from the actual slot so that it's clearer. Probably on the endless list of backlogs though.